File: Code/search_algorithms/dijkstra_search/dijkstra.js

// Dijkstra's algorithm implementation - done by @NeboCode on Github

// Sample dataGraph with node distances which we will use to implement Dijkstra's algorithm.

let DijkstraGraph = { startingPoint: {A: 5, B: 2}, A: {C: 1, D: 2}, B: {A: 8, D: 7}, C: {D: 2, finishPoint: 3},D: {finishPoint: 4}, E: {C:4, B:5, finishPoint:7}, F:{D:3, B:2,finishPoint:4}, finishPoint: {}};



// Function to find the node with the lowest weight, using a JS reducer.

function findLowestNode (weights, finishedNodes) {
    let knownNodes = Object.keys(weights)
    let lowestNode = knownNodes.reduce((lowest, node) => {
    if (lowest === null && !finishedNodes.includes(node)) {
     lowest = node;
     }
   if (weights[node] < weights[lowest] && !finishedNodes.includes(node)) {    
   lowest = node;
    }
   return lowest;
   }, null);
   
  return lowestNode
 };



// Dijkstra's algo implementation in JavaScript

function dijkstraAlgo(dataGraph) {

    let weights = Object.assign({finishPoint: Infinity}, dataGraph.startingPoint); 
    let parents = {finishPoint: null};  

    for (let child in dataGraph.startingPoint) {    
      parents[child] = 'startingPoint';  
     }

     // "caching" already visited nodes, so we don't come back to them!
    let finishedNodes = [];
    let node = findLowestNode(weights, finishedNodes);
       
    // finding the lowest weight, once the loop is done, we’ll have the lowest weight to reach the final point node
    while (node) {
    let weight = weights[node];
    let children = dataGraph[node]; 
    for (let n in children) {   
        let newWeight = weight + children[n];     
         if (!weights[n] || weights[n] > newWeight) { 
          weights[n] = newWeight; 
          parents[n] = node;
            }
         }
     finishedNodes.push(node);  
     node = findLowestNode(weights, finishedNodes);
    }
    let optimalPath = ['finishPoint'];
    let parent = parents.finishPoint;

while (parent) {
    optimalPath.unshift(parent);
    parent = parents[parent]; 
  }
  
  let result = { totalDistance: weights.finishPoint, optimalPath: optimalPath };

  return result;
};

File: Code/search_algorithms/dijkstra_search/dijkstra.py
start_node = 'start'
end_node = 'finish'

# Sample adjacency list of a Directed Acyclic Graph
graph = {
    start_node: {'A': 5, 'B': 2},
    'A': {'C': 1, 'D': 2},
    'B': {'A': 8, 'D': 7},
    'C': {'D': 2, end_node: 3},
    'D': {end_node: 4},
    'E': {'C': 4, 'B': 5, end_node:7},
    'F': {'D': 3, 'B': 2, end_node:4},
    end_node: {}
}


def find_lowest_node(weights, visited_nodes):
    known_nodes = weights.keys()
    lowest = None
    for node in known_nodes:
        if (lowest is None) and (node not in visited_nodes):
            lowest = node
        if (
            (lowest is not None) and (weights[node] < weights[lowest])
            and (node not in visited_nodes)
        ):
            lowest = node
    return lowest


def dijkstra(graph):
    weights = graph[start_node]  # start with children of start node
    weights[end_node] = float("inf")
    parents = {end_node: None}

    for child in graph[start_node]:
        parents[child] = start_node
    
    visited_nodes = []
    node = find_lowest_node(weights, visited_nodes)

    while node:
        weight = weights[node]
        children = graph[node]
        for n in children:
            new_weight = weight + children[n]
            if n not in weights:
                weights[n] = new_weight
                parents[n] = node
                continue
            if weights[n] > new_weight:
                weights[n] = new_weight
                parents[n] = node
        visited_nodes.append(node)
        node = find_lowest_node(weights, visited_nodes)
    optimal_path = [end_node]
    parent = parents[end_node]

    while parent:
        optimal_path.insert(0, parent)
        if parent in parents:
            parent = parents[parent]
        else:
            parent = None
    
    result = {
        'total_distance': weights[end_node],
        'optimal_path': optimal_path
    }

    return result
